compiler/testdata/coroutine_durable.go

```diff
@@ -3252,14 +3252,68 @@ func IdentityGenericInt(n int) { IdentityGeneric[int](n) }
 
 //go:noinline
 func IdentityGenericStructInt(n int) { (&IdentityGenericStruct[int]{n: n}).Run() }
+
+//go:noinline
+func IdentityGenericClosure[T any](_fn0 T) {
+	_c := coroutine.LoadContext[T, any]()
+	var _f0 *struct {
+		IP int
+		X0 T
+		X1 func()
+	} = coroutine.Push[struct {
+		IP int
+		X0 T
+		X1 func()
+	}](&_c.Stack)
+	if _f0.IP == 0 {
+		*_f0 = struct {
+			IP int
+			X0 T
+			X1 func()
+		}{X0: _fn0}
+	}
+	defer func() {
+		if !_c.Unwinding() {
+			coroutine.Pop(&_c.Stack)
+		}
+	}()
+	switch {
+	case _f0.IP < 2:
+		_f0.X1 = buildClosure(_f0.X0)
+		_f0.IP = 2
+		fallthrough
+	case _f0.IP < 3:
+		_f0.X1()
+		_f0.IP = 3
+		fallthrough
+	case _f0.IP < 4:
+		_f0.X1()
+	}
+}
+
+// TODO: add this go:noinline directive automatically
+//
+//go:noinline
+func buildClosure[T any](n T) func() {
+	return func() {
+		coroutine.Yield[T, any](n)
+	}
+}
+
+//go:noinline
+func IdentityGenericClosureInt(n int) { IdentityGenericClosure[int](n) }
 func init() {
+	_types.RegisterFunc[func()]("github.com/stealthrocket/coroutine/compiler/testdata.(*IdentityGenericStruct[go.shape.int]).Run")
```

Comment:
I'm curious about this, wouldn't the signature of the function be `func(uintptr)` to accept the generic type dictionary as first argument?

Reply:
I'm not sure, and I'll see if I can create a test case where it fails otherwise.

It's my understanding that when the Go compiler constructs a static call to a method it'll automatically insert the dictionary as the first argument (before the receiver). If you instead take a reference to the method so it can be called dynamically, it'll create a closure with the dictionary automatically bound to the function, and possibly the receiver too depending on how you've taken the reference. In neither case does the dictionary need to be explicitly listed as a parameter?

Reply:
Ah yeah it might not be needed, there's a parallel indeed with the receiver being passed in a register (DX), the dictionary is given in AX, that's just a modification of the calling convention but not a change in the Go type or function signature 👍